lib/cache.js

```diff
@@ -0,0 +1,184 @@
+
+/*jslint nomen:true, anon:true, node:true, esversion:6  */
+'use strict';
+
+/**
+ * Entry class used as map values and intrusive linked list nodes.
+ */
+class Entry {
+    constructor (key, value, setAt, expiresAt, groupId) {
+        this.next = null;
+        this.prev = null;
+        this.key = key;
+        this.value = value;
+        this.setAt = setAt;
+        this.expiresAt = expiresAt;
+        this.groupId = groupId;
+    }
+}
+
+/**
+ * LRU cache.
+ * Supported options are {max: int} which will set the max capacity of the cache.
+ */
+class ConfigCache {
+    constructor(options) {
+        options = options || {};
+        this.max = options.max;
+        if(!Number.isInteger(options.max) || options.max < 0) {
+            console.log('WARNING: no valid cache capacity given, defaulting to 100. %s', JSON.stringify(options));
+            this.max = 100;
+        }
+        if(this.max === 0) {
+            this.get = this.set = this.getTimeAware = this.setTimeAware = function(){};
+        }
+        this.size = 0;
+        this.map = new Map(); //key -> Entry
+        this.youngest = null;
+        this.oldest = null;
+    }
+
+    /**
+     * Set a cache entry.
+     * @param {string} key Key mapping to this value.
+     * @param {*} value Value to be cached.
+     * @param {number} groupId Id of the entry's group, used to lazily invalidate subsets of the cache.
+     */
+    set(key, value, groupId) {
+        var entry = this.map.get(key);
+        if(entry !== undefined) {
+            entry.value = value;
+            entry.groupId = groupId;
+            this._makeYoungest(entry);
+            return;
+        }
+        if(this.size === this.max) {
+            entry = this.oldest;
+            this.map.delete(entry.key);
+            entry.key = key;
+            entry.value = value;
+            entry.groupId = groupId;
+            this.map.set(key, entry);
+            this._makeYoungest(entry);
+            return;
+        }
+        entry = new Entry(key, value, 0, 0, groupId);
+        this.map.set(key, entry);
+        if(this.size === 0) {
+            this.youngest = entry;
+            this.oldest = entry;
+            this.size = 1;
+            return;
+        }
+        entry.next = this.youngest;
+        this.youngest.prev = entry;
+        this.youngest = entry;
+        this.size++;
+    }
+
+    /**
+     * Set a time aware cache entry.
+     * @param {string} key Key mapping to this value.
+     * @param {*} value Value to be cached.
+     * @param {number} now  Current time.
+     * @param {number} expiresAt Time at which entry will become stale.
+     * @param {number} groupId Id of the entry's group, used to lazily invalidate subsets of the cache.
+     */
+    setTimeAware(key, value, now, expiresAt, groupId) {
```

Comment:
the body of this function is mostly the same as the `set()`, besides setting `now` and `expiresAt` value.  How about creating a `_set()` function for the common logic, and have `set()` and `setTimeAware()` just call `_set()`.